File: companies/New Folder With Items/3882.minimum-xor-path-in-a-grid.go
package leetcode

//
// @lc app=leetcode id=3882 lang=golang
//
// [3882] Minimum XOR Path in a Grid
//
// Notes
// For each cell, keep the set of all XOR values reachable at that cell from the
// top or left. The DP array is one row wide: dp[col] is the previous-row set
// before update and the current-row set after update. Go map[int]struct{} is
// the set representation. Time: O(R*C*S), where S is reachable XOR count. Space:
// O(C*S).
//
// @lc code=start

func MinCost3882(grid [][]int) int {
	rows, cols := len(grid), len(grid[0])
	dp := make([]map[int]struct{}, cols)

	for row := 0; row < rows; row++ {
		for col := 0; col < cols; col++ {
			value := grid[row][col]
			if row == 0 && col == 0 {
				dp[col] = map[int]struct{}{value: {}}
				continue
			}

			reachable := map[int]struct{}{}
			if row > 0 {
				for previousXor := range dp[col] {
					reachable[previousXor^value] = struct{}{}
				}
			}
			if col > 0 {
				for previousXor := range dp[col-1] {
					reachable[previousXor^value] = struct{}{}
				}
			}
			dp[col] = reachable
		}
	}

	answer := 1 << 60
	for value := range dp[cols-1] {
		if value < answer {
			answer = value
		}
	}
	return answer
}

// @lc code=end

File: companies/Google/3882.minimum-xor-path-in-a-grid.java
/*
 * @lc app=leetcode id=3882 lang=java
 *
 * [3882] Minimum XOR Path in a Grid
 *
 * Standard grid DP, but each cell can have many reachable XOR values. dp[col]
 * stores the set of XORs reaching the current row's cell in that column; combine
 * from top and left, xor with the current value, and keep the final minimum.
 *
 * Time: O(rows * cols * states). Space: O(cols * states).
 */

import java.util.HashSet;
import java.util.Set;

// @lc code=start
class Solution {
    public int minCost(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        Set<Integer>[] dp = new HashSet[cols];
        for (int c = 0; c < cols; c++) {
            dp[c] = new HashSet<>();
        }

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                int value = grid[r][c];
                if (r == 0 && c == 0) {
                    dp[c] = new HashSet<>();
                    dp[c].add(value);
                    continue;
                }

                Set<Integer> reachable = new HashSet<>();
                if (r > 0) {
                    for (int previous : dp[c]) {
                        reachable.add(previous ^ value);
                    }
                }
                if (c > 0) {
                    for (int previous : dp[c - 1]) {
                        reachable.add(previous ^ value);
                    }
                }
                dp[c] = reachable;
            }
        }

        int answer = Integer.MAX_VALUE;
        for (int value : dp[cols - 1]) {
            answer = Math.min(answer, value);
        }
        return answer;
    }
}
// @lc code=end
